README
PkgConfig

This module is a pure perl version of pkg-config, a program
that provides compiler and linker flags for installed packages.
See:

http://www.freedesktop.org/wiki/Software/pkg-config/

for the original.  The POD for the module contains a more detailed
description of how it works and the differences from the original.
You can view it before installing using perldoc:

 % perldoc lib/PkgConfig.pm

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

You may change the scripts that get installed by using the --script
option one or more times.  For example, if you prefer not to install
pkg-config.pl you could do this:

 % perl Makefile.PL --script=ppkg-config

Or if you want to use ppkg-config and pkg-config (say you are on a
platform that  does not provide it):

 % perl Makefile.PL --script=pkg-config --script=ppkg-config

Or if you don't want any scripts installed at all:

 % perl Makefile.PL --script=none

The default is "--script=ppkg-config --script=pkg-config.pl" except for
Windows, where the default is just "--script=ppkg-config".

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc PkgConfig

You can also look for information at:

    GitHub, which hosts the projects official repository
        https://github.com/PerlPkgConfig/perl-PkgConfig
    
    Project's issue tracker on GitHub
        https://github.com/PerlPkgConfig/perl-PkgConfig/issues

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/PkgConfig

    CPAN Ratings
        http://cpanratings.perl.org/d/PkgConfig

    MetaCpan
        https://metacpan.org/pod/PkgConfig


LICENSE AND COPYRIGHT

Copyright (C) 2012 M. Nunberg

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for PkgConfig

  - Handle spaces in Perl and PkgConfig.pm path (gh#47, gh#52)

0.24026 2020-07-19 06:17:28 -0600
  - Fixed path on FreeBSD to use te modern defaults (gh#49, tomhukins++)

0.23026 2018-10-15 05:31:23 -0400
  - Fixed only used once warning (gh#45, gh#46 Ari Jolma++)

0.22026 2018-09-03 09:08:39 -0400
  - On OS X with Homebrew, be more aggresive in finding installed packages

0.21026 2018-08-15
  - promote to production

0.20026_01 2018-08-15
  - Fix bug where Strawberry Perl PAR::Packer packed PkgConfig.pm
    could die on startup (gh#45, gh#44 SLAFFAN++)

0.19026 2017-11-21
  - Fix requirements for optional Perl::MinimumVersion test

0.18026 2017-11-21
  - Add support for Slackware 64bit

0.17026 2017-09-18
  - --modversion does not require dependencies

0.16026 2017-03-23
  - Fix for Perl 5.25.11+ where '.' is not in @INC

0.15026 2017-01-26
  - New release with fixes to META.json

0.14026 2017-01-26
  - support for static cflags.  This has been supported
    by pkgconf for a while.

0.12026 2016-05-09
  - Updated project URLs in README
  - Spelling fixes in various documentation files
  - Support for MSYS2

0.11026 2016-01-25
  - Do not install misc .pl files in the distribution root.

0.10026 2016-01-08
  - promote to production

0.09026_02 2016-01-07
  - Test and development changes only.  No functional or
    bug fixes to PkgConfig.pm itself.

0.09026_01 2016-01-05
  - Remove use of flock
    It isn't portable and was causing hang/failures on NFS
  - Add --short-errors option

0.09026 2015-03-18
  - Add --errors-to-stdout option
  - Support for Gentoo Linux

0.08926 2015-03-05
  - Add --atleast-pkgconfig-version option

0.08826 2015-02-23
  - Use 64bit pkgconfig paths on 64 bit solaris

0.08726 2015-02-19
  - Parallel test if multiple CPUs can be detected (Linux, Windows)
  - Skip long running tests under ActiveState PPM build
    These were causing timeouts in dependencies such as Alien::Base

0.08626 2015-01-05
  - Add support for --libs-only-other option (GH#18 garu)
  - Add support for --cflags-only-other option (GH#17 garu)
  - Remove system multi arch library directory flags (example: -L/usr/lib/x86_64-linux-gnu) for Debian 7+
    (patches for other distributions are encouraged)

0.08526 2014-10-15
  - Add support for PKG_CONFIG_LIBDIR, PKG_CONFIG_ALLOW_SYSTEM_CFLAGS and PKG_CONFIG_ALLOW_SYSTEM_LIBS environment variables
  - Add support for windows registry keys
  - Use short (MSDOS 8.3) names on Windows for search paths that contain white space
  - Replace forward slash (/) in search paths where back slash is provided (\)

0.08420 2014-08-03
  - promote to production
  - Code named Salt Lake City

0.08320_03 2014-08-02
  - minor optimisations

0.08320_02 2014-08-01
  - get_cflags returns a string in scalar context
  - get_ldflags returns a string in scalar context
  - ppkg-config program correctly escapes spaces

0.08320_01 2014-07-31
  - handle quoted space

0.08320 2014-07-16
  - promote to production

0.08220_02 2014-07-16
  - fix regression in last revision:
    error in dependent component is propagated to main component

0.08220_01 2014-07-15
  - fixed bug where dependent components were having their variables
    overridden by the main .pc file

0.08220 2014-06-25
  - promote to production

0.08120_01 2014-06-24
  - better Debian detection and support
  - silence warning on Debian 6 or earlier when
    dpkg-dev is installed

0.08120 2014-06-20
  - require ExtUtils::MakeMaker 6.56 for BUILD_REQUIRES

0.08020 2014-06-17
  - promote to production

0.07920_01 2014-06-16
  - add support for the --cflags-only-I option

0.07920 2014-06-13
  - require Test::More >= 0.94 for subtest
    reading comprehension is key

0.07820 2014-06-13
  - require Test::More > 0.90 and != 0.92 for subtest

0.07720 2014-06-09
  - promote to production
  - added PERL_PKG_CONFIG_SCRIPTS environment variable

0.07620_01 2014-06-08
  - added get_var public accessor method
  - support for constraints =, !=, >=, <=, >, and <

0.07620 2014-06-06
  - promote to production

0.07520_04 2014-06-03
  - windows test fix

0.07520_03 2014-06-03
  - fix split on Makefile.PL --script option

0.07520_02 2014-06-03
  - make appropriate exclusions for Strawberry Perl

0.07520_01 2014-06-03
  - support for ${pcfiledir}
  - base strawberry root on $^X instead of Config.pm location

0.07520 2014-06-02
  - work around for Strawberry Perl 5.20.0.1

0.07420 2014-05-09
  - add --script option for Makefile.PL (GH#10 kmx)
  - add --list-all option for ppkg-config script (GH#10 kmx)

0.07320 2014-05-06
  - Use .pc files from Strawberry Perl on MSWin32 + Strawberry Perl

0.07220 2014-04-26
  - fix possible 6.2 incompatibility
  - remove unintentional new line from default paths on Debian

0.07120 2014-03-26
  - documentation tweak to improve Kwalitee score

0.07020 2014-03-25
  - use the same search path as the pkg-config normally used on these
    systems:
    
      Debian Linux
      Debian kFreeBSD
      RedHat/CentOS/Fedora Linux
      FreeBSD*
      NetBSD
      OpenBSD
      
      * FreeBSD 10 doesn't support PKG_CONFIG_DEBUG_SPEW so I can't verify
        the path there, but the path is correct for FreeBSD 9
        
    Non .deb or .rpm based Linux distributions are not detected and will 
    use the default non-platform specific search path.  .rpm based 
    distributions other than RedHat/CentOS/Fedora may be misidentified 
    as one of those.  Patches to correct unidentified or misidentified 
    platforms would be eagerly received.
    
    this is a significant change in behavior, but arguably a useful one
  - add file_path option for loading specific .pc file

0.06420 2014-02-13
  - use ; instead of : for PKG_CONFIG_PATH on MSWin32
  - support --exact-version, --atleast-version and --max-version
    (at least enough to support ExtUtils::PkgConfig)
  - Fix Libs.private property and therefore --static option which
    was being hidden due to a bug

test.pl
use strict;
use warnings;
use blib;
use Test::Harness;
use File::Spec;
use File::Glob qw( bsd_glob );
use lib map { File::Spec->rel2abs($_) } qw( blib/lib blib/arch );

eval {
  require PkgConfig;
  print "\nTesting PkgConfig $PkgConfig::VERSION, Perl $], $^X\n";
};

my $cpu_count;

if(defined $ENV{NUMBER_OF_PROCESSORS})
{
  $cpu_count = $ENV{NUMBER_OF_PROCESSORS};
}
elsif(-r "/proc/cpuinfo")
{
  my $fh;
  open $fh, '<', '/proc/cpuinfo';
  $cpu_count = scalar grep /^processor\s/, <$fh>;
  close $fh;
}

if(defined $cpu_count && $cpu_count > 0)
{
  $cpu_count = 8 if $cpu_count > 8;
  print "Testing on $cpu_count CPUS\n";
  $ENV{HARNESS_OPTIONS} = "j$cpu_count";
}

Test::Harness::runtests(bsd_glob 't/*.t');


gen_multi_tests.sh
#!/bin/sh
cd t
templ=iterfiles_template
test_base="02-iterfiles"
find usr | split -l 50 - FLIST

for f in FLIST*; do
    cp $templ "$test_base-$f.t"
done


t/PkgConfigTest.pm
#!perl
package PkgConfigTest;
use strict;
use warnings;
use Config;
use Test::More;
use Data::Dumper;
use File::Spec;
use File::Basename qw(fileparse);
use Config;
use FindBin ();
use Exporter;
our @ISA = qw( Exporter );

$ENV{PKG_CONFIG_NO_OS_CUSTOMIZATION} = 1;

our @EXPORT = qw(
    expect_flags run_common $RV $S);

my @PC_PATHS = qw(data/usr/lib/pkgconfig data/usr/share/pkgconfig
                data/usr/local/lib/pkgconfig data/usr/local/share/pkgconfig);
                


my $TARBALL = File::Spec->catfile($FindBin::Bin, 'pc_files.tar.gz');
my $LOCK    = File::Spec->catfile($FindBin::Bin, 'pc_files.lock');
@PC_PATHS = map { $FindBin::Bin . "/$_" } @PC_PATHS;
@PC_PATHS = map {
    my @components = split(/\//, $_);
    $_ = File::Spec->catfile(@components);
    $_;
} @PC_PATHS;
    
note Dumper(\@PC_PATHS);

$ENV{PKG_CONFIG_PATH} = join($Config{path_sep}, @PC_PATHS);

our $RV;
our $S;

my $SCRIPT = "$FindBin::Bin/../script/ppkg-config";

# Work around git on windows' lamentable lack of symbolic
# link support
$SCRIPT = $FindBin::Bin . "/../lib/PkgConfig.pm"
    if $^O eq 'MSWin32' && -d '.git';

sub run_common {
    my @args = @_;
    (my $ret = qx($^X $SCRIPT --env-only @args))
        =~ s/(?:^\s+)|($?:\s+$)//g;
    $RV = $?;
    $S = $ret;
}

sub expect_flags {
    my ($flags,$msg) = @_;
    like($S, qr/\Q$flags\E/, $msg);
}

sub run_exists_test {
    my ($flist,$pmfile) = @_;
    note "$pmfile: Will perform --exist tests";
    foreach my $fname (@$flist) {
        next unless -f $fname;
        my ($base) = fileparse($fname, ".pc");
        run_common("$base");
        ok($RV == 0, "Package $base exists");
    }
}

sub _single_flags_test {
    my $fname = shift;
    return unless -f $fname;
    my ($base) = fileparse($fname, ".pc");
    run_common("--libs --cflags $base --define-variable=prefix=blah");
    ok($RV == 0, "Got OK for --libs and --cflags");
    if($S =~ /-(?:L|I)/) {
        if($S !~ /blah/) {
            
            #these files define $prefix, but don't actually use them for
            #flags:
            if($base =~ /^(?:glu?|libconfig)$/) {
                note "Skipping gl pcfiles which define but do not use 'prefix'";
                return;
            }
            
            #Check the file, see if it at all has a '$prefix'
            open my $fh, "<", $fname;
            if(!defined $fh) {
                note "$fname: $!";
                return;
            }
            
            my @lines = <$fh>;
            if(grep /\$\{prefix\}/, @lines) {
                ok(0, "Expected substituted prefix for $base");
            } else {
                note "File $fname has no \${prefix} directive";
            }
            return;
        }
        ok($S =~ /blah/, "Found modified prefix for $base");
    }
}

sub run_flags_test {
    my ($flist,$pmfile) = @_;
    note "$pmfile: Will perform --prefix, --cflags, and --libs tests";
    foreach my $fname (@$flist) {
        _single_flags_test($fname);
    }
}

1;


win32/fixlinks.pl
use strict;
use warnings;
use File::Spec;

# this script requires administrator privileges

warn "this script requires admin privilages, which you don't appear to have"
  unless eval { require Win32; Win32::IsAdminUser(); 1 };

my @list = map { chomp; [split /\t/]->[1] } grep /^120000/, `git ls-files -s `;

foreach my $link (@list)
{
  # in case we ran this before
  system "git checkout $link";
  
  my $target = do { open my $fh, '<', $link; local $/; <$fh> };
  unlink $link;

  my $cmd = "mklink " . File::Spec->catfile($link) . " " . File::Spec->catfile($target); 
  print "> $cmd\n";
  system $cmd;
  
  if($?)
  {
    warn "failed";
    # revert on failure
    system "git checkout $link";
    next;
  }
  
  # ignore change so that it won't be comitted back
  print "> git update-index --assume-unchanged $link\n";
  system 'git', 'update-index', '--assume-unchanged', $link;
}

print "press <ENTER> to continue.\n";
<STDIN>;


win32/fix_links.pl
use strict;
use warnings;
use File::Spec;

# this script requires administrator privileges

warn "this script requires admin privilages, which you don't appear to have"
  unless eval { require Win32; Win32::IsAdminUser() };

my @list = map { chomp; [split /\t/]->[1] } grep /^120000/, `git ls-files -s `;

foreach my $link (@list)
{
  # in case we ran this before
  system "git checkout $link";
  
  my $target = do { open my $fh, '<', $link; local $/; <$fh> };
  unlink $link;

  my $cmd = "mklink " . File::Spec->catfile($link) . " " . File::Spec->catfile($target); 
  print "> $cmd\n";
  system $cmd;
  
  if($?)
  {
    warn "failed";
    # revert on failure
    system "git checkout $link";
    next;
  }
  
  # ignore change so that it won't be comitted back
  print "> git update-index --assume-unchanged $link\n";
  system 'git', 'update-index', '--assume-unchanged', $link;
}

print "press <ENTER> to continue.\n";
<STDIN>;


Makefile.PL
use strict;
use warnings;
use 5.006;
use ExtUtils::MakeMaker 6.56;
use Getopt::Long;
use Config;

### Makefile.PL commandline params
# e.g. Makefile.PL --script=ppkg-config --script=pkg-config
Getopt::Long::GetOptions("script=s" => \my @scripts);

if (@scripts == 0 && defined $ENV{PERL_PKG_CONFIG_SCRIPTS})
{
  @scripts = ($ENV{PERL_PKG_CONFIG_SCRIPTS});
}

if (@scripts) {
  @scripts = map { split /,/ } @scripts;
  if (grep /^none$/, @scripts) {
    @scripts = ();
  } else {
    @scripts = map { "script/$_" } @scripts;
    -f $_ or die "non-existing '$_'" for (@scripts);
  }
}
else {
  @scripts = ( 'script/ppkg-config', $^O ne 'MSWin32' ? 'script/pkg-config.pl' : () );
  if($^O eq 'MSWin32' && $Config{myuname} =~ /strawberry-perl/) {
    if(eval q{ require 5.020 }) {
      push @scripts, 'script/pkg-config';
    
      print "\n\n";
      print "!!! Note that script will be installed as pkg-config  !!!\n";
      print "!!! in addition to ppkg-config since you are using    !!!\n";
      print "!!! Strawberry Perl 5.20 or better.  This is probably !!!\n";
      print "!!! What you want                                     !!!\n";
      print "\n\n";
      
      if($Config{myuname} =~ /strawberry-perl 5\.20\.0\.1/)
      {
        print "\n\n";
        print "!!! Some of the .pc files that come with your version !!!\n";
        print "!!! of Strawberry Perl may not work correctly with    !!!\n";
        print "!!! this version of PkgConfig.  You should consider   !!!\n";
        print "!!! patching them.  See README.win32 for details.     !!!\n";
        print "\n\n";
      }
      
    } else {
      print "\n\n";
      print "!!! You may want to consider installing the script as !!!\n";
      print "!!! pkg-config as it can then be used by              !!!\n";
      print "!!! ExtUtils::PkgConfig (This is how it works in      !!!\n";
      print "!!! Strawberry 5.20 and better in fact).  Please see  !!!\n";
      print "!!! details in README.win32 if you are interested.    !!!\n";
      print "\n\n";
    }
  }
}

WriteMakefile(
    NAME                => 'PkgConfig',
    AUTHOR              => q{M. Nunberg <mnunberg@haskalah.org>},
    VERSION_FROM        => 'lib/PkgConfig.pm',
    ABSTRACT_FROM       => 'lib/PkgConfig.pm',
    LICENSE             => 'perl',
    MIN_PERL_VERSION    => '5.006000',
    PL_FILES            => {},
    EXE_FILES           => \@scripts,
    PREREQ_PM => {
        'Test::More' => '0.94',
        'Env' => '1.04',
    },
    BUILD_REQUIRES => {
        'ExtUtils::MakeMaker' => '6.56',
    },
    PM => {
      "lib/PkgConfig.pm" => "\$(INST_LIB)/PkgConfig.pm",
    },
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            bugtracker => {
                web  => 'https://github.com/PerlPkgConfig/perl-PkgConfig/issues',
            },
            homepage => 'https://metacpan.org/pod/PkgConfig',
            repository => {
                type => 'git',
                url  => 'https://github.com/PerlPkgConfig/perl-PkgConfig.git',
                web  => 'https://github.com/PerlPkgConfig/perl-PkgConfig',
            },
            x_IRC => "irc://irc.perl.org/#native",
        },
        provides => {
            PkgConfig => {
                file    => "lib/PkgConfig.pm",
                version => do { require "./lib/PkgConfig.pm"; $PkgConfig::VERSION; },
            },
        },
    },
    
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean               => { FILES => 'PkgConfig-*' },
);

sub MY::test_via_harness
{
  my($self, $perl, $tests) = @_;
  "\t$perl run_test.pl"
}

